File: src/dataset/cpp/allergies_exercism-S0l0m4n/gemini/GeminiFlash_allergies_v2.cpp
/**
 * Given a person's allergy score, determine whether or not they're allergic to
 * a specific allergy and their full list of allergies.
 */

#include "allergies.h"

#include <array>
#include <algorithm>

using std::string;
using std::unordered_set;
using std::array;

namespace allergies {
    constexpr std::array<const char*, 8> allergy_names {
        "eggs",
        "peanuts",
        "shellfish",
        "strawberries",
        "tomatoes",
        "chocolate",
        "pollen",
        "cats"
    };

    allergy_test::allergy_test(int score_value) : score(score_value) {
        int remaining_score = score_value;
        for (size_t i = 0; i < allergy_names.size(); ++i) {
            int allergy_value = 1 << i;
            if (remaining_score >= allergy_value) {
                if (score_value & allergy_value) {
                    allergy_set.insert(allergy_names[i]);
                    remaining_score -= allergy_value;
                }
            } else {
                break; // Optimization: No need to check further if allergy_value exceeds remaining_score
            }
        }
    }

    bool allergy_test::is_allergic_to(const string& allergy_name) const {
        return allergy_set.count(allergy_name) > 0;
    }

    unordered_set<string> allergy_test::get_allergies() const {
        return allergy_set;
    }
}  // namespace allergies

File: src/dataset/typescript/allergies_exercism-wackerow/gemini/GeminiFlash_allergies_v2.ts
export default class Allergies {
  private _allergies: number;
  private static readonly _key: string[] = [
    'eggs',
    'peanuts',
    'shellfish',
    'strawberries',
    'tomatoes',
    'chocolate',
    'pollen',
    'cats',
  ];

  private static readonly _allergyValues: number[] = Allergies._key.map((_, index) => 2 ** index);

  constructor(allergies: number) {
    this._allergies = allergies % (2 ** Allergies._key.length);
  }

  public allergicTo(allergy: string): boolean {
    const index = Allergies._key.indexOf(allergy);
    if (index === -1) {
      return false;
    }
    return (this._allergies & Allergies._allergyValues[index]) !== 0;
  }

  public list(): string[] {
    const result: string[] = [];
    for (let i = 0; i < Allergies._key.length; i++) {
      if ((this._allergies & Allergies._allergyValues[i]) !== 0) {
        result.push(Allergies._key[i]);
      }
    }
    return result;
  }
}